<p>Java里 <code>throws</code> 声明中的异常是多余的，如果它是:</p>
<ul>
  <li> 列出多次 </li>
  <li> 另一个列出的异常的子类（subclass）</li>
  <li> <code>RuntimeException</code>，或其子类（descendant）之一 </li>
  <li> 完全没必要，因为实际上不能抛出声明的异常类型 </li>
</ul>
<h2>不合规的代码示例</h2>
<pre>
void foo() throws MyException, MyException {}  // 不合规； 应列出一次
void bar() throws Throwable, Exception {}  // 不合规； Exception 是 Throwable 的子类
void baz() throws RuntimeException {}  // 不合规； 始终可以抛出RuntimeException 
</pre>
<h2>合规的解决方案</h2>
<pre>
void foo() throws MyException {}
void bar() throws Throwable {}
void baz() {}
</pre>
<h2>例外</h2>
<p>对于无法从方法体（method body）中抛出的异常，该规则不会报告任何问题:</p>
<ul>
  <li> 在重写和实现方法中 </li>
  <li> 在接口 <code>default</code> 方法中 </li>
  <li> 在非私有方法中只有 <code>throw</code>， 有空的方法体或单个return语句。 </li>
  <li> 在可覆盖的方法中（非final，或不是final类的成员，非静态，非私有），如果使用恰当的javadoc记录异常。 </li>
</ul>
<pre>
class A extends B {
  @Override
  void doSomething() throws IOException {
    compute(a);
  }

  public void foo() throws IOException {}

  protected void bar() throws IOException {
    throw new UnsupportedOperationException("This method should be implemented in subclasses");
  }

  Object foobar(String s) throws IOException {
    return null;
  }

  /**
   * @throws IOException Overriding classes may throw this exception if they print values into a file
   */
  protected void print() throws IOException { // 没问题，方法可重写而且异常有恰当的javadoc记录
    System.out.println("foo");
  }
}
</pre>

